connection-limit/connection_limit_kern.c

```diff
@@ -269,22 +269,34 @@ int _xdp_limit_conn(struct xdp_md *ctx)
     void *data_end = (void *)(long)ctx->data_end;
     void *data = (void *)(long)ctx->data;
 
-    /* Check if its valid ethernet packet */
-    if (data + sizeof(struct ethhdr)+ 1 > data_end)
+    struct ethhdr *eth = data;
+
+    /* Check if it is a valid ethernet packet */
+    if (data + sizeof(*eth)+ 1 > data_end)
         return XDP_PASS;
+   
+    uint16_t eth_type = eth->h_proto;
 
-    /* Check if its valid ip packet */
-    struct iphdr *iph = (struct iphdr *)(data + sizeof(struct ethhdr));
-    if (iph + 1 > data_end)
+    /* Ignore other than IPV4 packets */
+    if (ntohs(eth_type) != ETH_P_IP)
+        return XDP_PASS;
+
+    struct iphdr *iph = (struct iphdr *)(data + sizeof(*eth));
+    __u8 l4_offset = iph->ihl * 4; // ipv4 header length
+
+    /* Check if it is a valid IPV4 packet */
+    if (iph->ihl < 5 || ((unsigned char *)iph + l4_offset) > data_end)
```

Comment:
Range of IHL is 5 to 15, with boundary values included. Please add check explicitly for this.